src/main/kotlin/org/rust/lang/utils/snapshot/SnapshotList.kt

```diff
@@ -0,0 +1,27 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.utils.snapshot
+
+import org.rust.stdext.removeLast
+
+class SnapshotList<E> private constructor(
+    private val inner: MutableList<E>
+) : Snapshotable(), Collection<E> by inner {
```

Comment:
There is a leak. E.g if you remove an element from a list, this action will not be logged. Consider not implementing any standard interfaces an just write methods you need by yourself. Also make `inner` really "inner" without possibility to set it externally (like in SnapshotMap).
Same with SnapshotSet.